# EX_2

손글씨, 와인 분류, 유방암 진단

In [2]:
#데이터 불러오기
from sklearn.datasets import load_digits
from sklearn.datasets import load_wine
from sklearn.datasets import load_breast_cancer
#학습 및 결과에 필요한 패키지 불러오기
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
#적용할 모델 불러오기
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

데이터 준비

In [79]:
#손글씨
digits = load_digits()
digits_data = digits.data
digits_label = digits.target

#와인
wine = load_wine()
wine_data = wine.data
wine_label = wine.target

#유방암
breast_cancer = load_breast_cancer()
breast_cancer_data = b_cancer.data
breast_cancer_label = b_cancer.target

## 1. 손글씨 분류

In [36]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,  
                                                    digits_label,  
                                                    test_size=0.2,   
                                                    random_state=7)
                                                   

In [37]:
#Decision Tree

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



In [38]:
#Random Forest

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



In [39]:
#SVM

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [40]:
#SGD Classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.83      0.95      0.89        42
           2       1.00      1.00      1.00        40
           3       1.00      0.82      0.90        34
           4       0.88      1.00      0.94        37
           5       0.78      1.00      0.88        28
           6       1.00      0.93      0.96        28
           7       0.97      0.91      0.94        33
           8       0.88      0.86      0.87        43
           9       0.96      0.75      0.84        32

    accuracy                           0.92       360
   macro avg       0.93      0.92      0.92       360
weighted avg       0.93      0.92      0.92       360



In [41]:
#Logistic Regression

logistic_model = LogisticRegression(max_iter=3000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.92      0.81      0.86        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



## 손글씨 분류 모델평가

모델 | 정확도 | precision | recall   
:---: | :---: | :---: | :---:  
Decision Tree       | 0.86 | 0.86 | 0.86
Random Forest     | 0.96  | 0.96 | 0.96
SVM               | 0.99  | 0.99 | 0.99
SGD Classifier    | 0.93  | 0.93 | 0.93
 Logistic Regression | 0.95| 0.95 | 0.95

---

모델별 정확도와 precision, recall 평균값을 비교해 봤을 때,
SVM모델이 모든 지표에서 가장 높게 나왔으므로 가장 적절한 선택이 되겠다.

## 2. 와인 분류

In [42]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data,  
                                                    wine_label,  
                                                    test_size=0.2,   
                                                    random_state=7)

In [43]:
#Decision Tree

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



In [44]:
#Random Forest

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [45]:
#SVM

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36



In [53]:
#SGD Classifier

sgd_model = SGDClassifier(max_iter=3000)
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      1.00      0.93         7
           1       0.88      0.41      0.56        17
           2       0.55      0.92      0.69        12

    accuracy                           0.69        36
   macro avg       0.77      0.78      0.73        36
weighted avg       0.77      0.69      0.68        36



In [62]:
#Logistic Regression

logistic_model = LogisticRegression(max_iter=4000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.94      1.00      0.97        17
           2       1.00      0.92      0.96        12

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



## 와인 분류 모델평가


모델 | 정확도 | precision | recall   
:---: | :---: | :---: | :---:  
Decision Tree       | 0.96 | 0.94 | 0.94
Random Forest     | 1.00  | 1.00 | 1.00
SVM               | 0.59  | 0.61| 0.56
SGD Classifier    | 0.77  | 0.78 | 0.73
 Logistic Regression | 0.98| 0.97 | 0.97

---

모델별 정확도와 precision, recall 비교해봤을 때 Random Forest가 모든 부분이 1.00으로
가장 높게 나왔으므로 가장 적절한 선택이 되겠다.

## 3. 유방암 진단

In [81]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,
                                                    breast_cancer_label,
                                                    test_size=0.2,   
                                                    random_state=7)

In [82]:
#Decision Tree

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114



In [83]:
#Random Forest

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114



In [84]:
#SVM

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



In [85]:
#SGD Classifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.88      0.83        40
           1       0.93      0.88      0.90        74

    accuracy                           0.88       114
   macro avg       0.86      0.88      0.87       114
weighted avg       0.88      0.88      0.88       114



In [87]:
#Logistic Regression

logistic_model = LogisticRegression(max_iter=3000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



## 유방암 진단 모델평가

모델 | 정확도 | precision | recall   
:---: | :---: | :---: | :---:  
Decision Tree       | 0.91 | 0.89 | 0.91
Random Forest     | 1.00  | 1.00 | 1.00
SVM               | 0.94  | 0.86| 0.90
SGD Classifier    | 0.86  | 0.88 | 0.88
 Logistic Regression | 0.96| 0.93 | 0.95
 
 ---
 모델별 정확도와 precision, recall 비교해봤을 때 Random Forest가 모든 부분이 1.00으로
가장 높게 나왔으므로 가장 적절한 선택이 되겠다.

## Exproation_2 회고

첫번쨰 익스였던 가위바위보 를 할때보다는 수월하게 했다. 가위바위보를 할 때는 모든 것을 다 이해하려고 해서 더 어렵게 느꼈던 거 같았다. 모르는 코드들이 너무나 많았기에 그것들에 대해서 찾아보고 이해하는데 시간이 너무 오래 걸렸다. 결국 다 이해하지도 못했다.
그래서 이번에는 그냥 흐름만 파악하려고 하였더니 수월하게 진행이 되었다. 코드에 대해서는 이론적으로 이해는 다 못했지만 이러한 동작을 하는구나 정도로 생각하고 넘어갔다. 이런 방식으로 하는게 맞는건지 가위바위보를 했던 거처럼 하는게 맞는건지 모르겠지만 익스를 진행하는 방식은 이번에 했던 방식이 나에게는 맞는거 같다. 코드의 용도만 알고 이론적인 부분은 나중에 이해해보는게 더 나을거 같다.